nixos/modules/programs/wayland/niri.nix

```diff
@@ -0,0 +1,73 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.programs.niri;
+in
+{
+  options.programs.niri = {
+    enable = lib.mkEnableOption "Niri, a scrollable-tiling Wayland compositor";
+
+    package = lib.mkPackageOption pkgs "niri" { };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+    portalPackage = lib.mkPackageOption pkgs "xdg-desktop-portal-gnome" { };
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      {
+        environment = {
+          systemPackages = [ cfg.package ];
+        };
+
+        services = {
+          displayManager.sessionPackages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+          gnome.gnome-keyring.enable = lib.mkDefault true;
+        };
+
+        systemd = {
+          packages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#authentication-agent
+          user.services.niri-polkit-agent = lib.mkIf config.security.polkit.enable {
+            description = "Niri PolicyKit Authentication Agent";
+            partOf = [ "graphical-session.target" ];
+            wantedBy = [ "niri.service" ];
+            serviceConfig = {
+              Type = "simple";
+              ExecStart = "${pkgs.polkit-kde-agent}/libexec/polkit-kde-authentication-agent-1";
+              BusName = "org.kde.polkit-kde-authentication-agent-1";
+              Restart = "on-failure";
+              Slice = "background.slice";
+              TimeoutSec = "5sec";
+            };
+          };
+        };
+
+        xdg.portal = {
+          enable = lib.mkDefault true;
+
+          configPackages = [ cfg.package ];
+          extraPortals = [ cfg.portalPackage ];
```

Comment:
Unrealted to the question of "should we enable this conditionally", it might be nicer to write it like so:

```suggestion
          extraPortals = lib.mkIf cfg.package.enableScreencastSupport [ cfg.portalPackage ];
```

###### actually we probably need to change the package to make this accessible on the result like that, don't we?

I didn't do that in my flake, because it needs to still support `nixos-24.05` which has a niri package that doesn't have this attribute, but we could reasonably do it here. I think it would be neat to only depend on the rust features to probe here, as that allows a user to add a custom `buildRustPackage` derivation *without* needing to conform to what nixpkgs does. But, in practice, a user is only ever likely to pass one of three different packages, all of which have this attribute as one of their parameters ([`github:NixOS/nixpkgs`](https://github.com/NixOS/nixpkgs/blob/b4f1a9af5f0801c31d38cdba2a5aa1e6ce3f6a58/pkgs/by-name/ni/niri/package.nix#L22), [`github:sodiboo/niri-flake`](https://github.com/sodiboo/niri-flake/blob/0a50b261ac5b6cdcc8eb7921a4f022c92b29d1ab/flake.nix#L82), [`github:YaLTeR/niri`](https://github.com/YaLTeR/niri/blob/d7156df8428cb9e6665a212e66c74bdb25842be8/flake.nix#L130))

the bottom line is that, the difference between these two ways of detecting screencast support is mostly a matter of taste. `niri-flake` has an actual limitation that forces it to do what it does, but we could do either. that is, if we even should do it at all, which i'm actually unsure about (hence asking YaLTeR). i personally prefer the first way i proposed, as it relies on cargo attributes which are present even on 24.05.

Reply:
Hey! I answered more in depth on Matrix; xdp-gnome isn't really required for anything other than the `xdp-gnome-screencast` feature (though, it may still be nice to have at the user's choice for things like the file chooser).

Reply:
I've kept the option but left a note about the GNOME portal being required for screencast support